Import scale transforms and physical unit support (finishes v2.1.0)

oanegros opened the pull request:
- Adds physical unit reading of files #92 #66 #61 
- defines editable setting of import scaling and import location 
- links this to update_settings in reload

also:
- refactors input parsing
- adds line thickness to axis mod panel

BradyAJohnston commented:
it's a real mess that's for sure, but this is a working solution (for during development). 

Now that Blender's add-ons are all extensions and they are stored in different repos, the `__package__` is different depending on where they are installed.

For a regular user who has installed from the extensions website, this will be `bl_ext.user_default.microscopynodes`, but for us during development this will likely be `bl_ext.vscode_development.molecularnodes`. 

I'm not entirely sure why the `__package__` causes issues during development, but having a function that tries both as a backup was my current fix. I don't like it (and maybe we should be trying some other key combos as well?) but from what I understand this should only be a development issue and not bother uers.

oanegros commented:
@BradyAJohnston Ah yes it read like something like this. I'm just mostly confused because during testing, microscopy nodes does not show up in the `context.preferences.addons` at all. 

Is there some step where you're explicitly installing the addon in blender in the test pipeline?

addons list : `[<bpy_struct, Addon("io_anim_bvh") at 0x1114ced08>, <bpy_struct, Addon("io_curve_svg") at 0x1114a6d78>, <bpy_struct, Addon("io_mesh_uv_layout") at 0x174339bc8>, <bpy_struct, Addon("io_scene_fbx") at 0x1114adb78>, <bpy_struct, Addon("io_scene_gltf2") at 0x111489c88>, <bpy_struct, Addon("cycles") at 0x13f2391d8>, <bpy_struct, Addon("pose_library") at 0x11149c358>, <bpy_struct, Addon("bl_pkg") at 0x174338f38>]`

BradyAJohnston commented:
For testing I do run a specific `_test_register()` which is mostly just the same as `register()`: https://github.com/BradyAJohnston/MolecularNodes/blob/37aa3da2c9ee6b5d045357af74c52d1a95f7e0b7/tests/conftest.py#L9

oanegros commented:
I am doing a similar  [test_register](https://github.com/oanegros/MicroscopyNodes/blob/055b95bc67b2a543b624830bd71759056e149dd6/tests/utils.py#L19)
but this seems to not add the package to the addons of the bpy system - it might be an issue with how i'm installing Blender/bpy in the workflow? I'll debug further

BradyAJohnston commented:
I'm not sure about the particulars of `pytest`, but using `conftest.py` ensures that everything in that is run before any of the tests start which seems to make sure it works for me.

oanegros commented:
Okay I've figured out why this is failing. 
I'm currently still running all my tests in `bpy` without a full Blender instance, and omitting the `pyopenvdb` tests. 
The addon-preferences only load within a Blender instance. 

I could reimplement the tests to handle this, but this makes microscopy nodes only available to run with blender in the `blender -b -p python_script.py`, where this is currently possible with only bpy and microscopynodes installed. Given that `pyopenvdb` is now bundled in bpy 4.4, i do think a fully separately installable `microscopynodes` PyPI/conda package would be ideal. 

My proposal:
make a `headless_preferences.json` that lists all addon preferences and can be manually changed if necessary, and is only loaded if addon preferences cannot be loaded.

oanegros commented:
Related to #84

BradyAJohnston commented:
If you want to start doing GHA tests using a blender installation, I've setup an action to install and alias Blender for running tests in GHA, which you can see in use here:

```yml
        steps:
            - uses: actions/checkout@v4
            - uses: BradyAJohnston/setup-blender@v3
              with:
                version: ${{ matrix.version }}
            - name: Install in Blender
              run: |
                blender -b -P tests/python.py -- -m pip install ".[test]"
            - name: Run Tests
              run: |
                blender -b -P tests/run.py -- -vv tests --cov --cov-report=xml
```

https://github.com/BradyAJohnston/MolecularNodes/blob/604d3fc097a2d8847b345eebd19f8cd0e6dd8769/.github/workflows/tests.yml#L18-L28

Once 4.4 is on `pypi` it should help a bit as well, as pytest no longer exits with a failed exit code even with all test success (there was always a segfault on exit), so 4.4 should help us a lot with running tests.

oanegros commented:
Ah interesting that the segfaulting is fixed! 
I had already implemented a workaround as the behaviour was linked to the callback function of properties and this was preventable with explicit deleting of callback attributes in `conftest`:

```
@pytest.hookimpl(trylast=True)
def pytest_sessionfinish(session, exitstatus):
    import microscopynodes
    # regrettably necessary, pytest segfaults if properties
    # with callback functions stay alive

    UPDATE_PROPS = [
        'MiN_input_file',
        'MiN_explicit_cache_dir',
        'MiN_selected_cache_option',
        'MiN_axes_order',
        'MiN_selected_zarr_level',
        'MiN_channel_nr',
        'MiN_reload',
    ]
    deleted = 0
    for prop in UPDATE_PROPS:
        try:
            delattr(bpy.types.Scene, prop)
            deleted += 1
        except:
            print(f"{prop} not found")

    microscopynodes.unregister()
    print(f'called session finish, {deleted} properties deleted')
    # raise ValueError
```

oanegros commented:
But for this PR, i think it works to rework the addon preferences to something workable, loading from file, and then implement cli overrides for this file/its settings later in a #84 full API reworking. But to not lock the program out from a full `bpy` implementation, i don't think i should move to testing in a full blender instance.

BradyAJohnston commented:
I think it would still be worth having at least some tests that run inside of Blender, as there still might be some differences between running inside of Blender and `bpy` as a module.

BradyAJohnston commented:
(you can also use `daily` to test against the latest daily build to catch and breaking API changes as they happen)

oanegros commented:
Right 😅 very good point! I created #100 to track this
